diff -rupN jellyfin-10.7.6.orig/MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs jellyfin-10.7.6/MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs
--- jellyfin-10.7.6.orig/MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs	2021-06-25 16:02:30.858817347 +0200
+++ jellyfin-10.7.6/MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs	2021-06-25 16:20:20.744192274 +0200
@@ -2677,9 +2677,9 @@ namespace MediaBrowser.Controller.MediaE
                 filters.Add("hwupload");
             }
 
-            // When burning in graphical subtitles using overlay_qsv, upload videostream to the same qsv context.
-            else if (isLinux && hasGraphicalSubs && (isQsvH264Encoder || isQsvHevcEncoder)
-                     && !(isTonemappingSupportedOnQsv && isVppTonemappingSupported))
+            // When burning in graphical subtitles using overlay_qsv or software decoding, upload videostream to the same qsv context.
+            else if (((isLinux && hasGraphicalSubs && !(isTonemappingSupportedOnQsv && isVppTonemappingSupported))
+                || isSwDecoder) && (isQsvH264Encoder || isQsvHevcEncoder))
             {
                 filters.Add("hwupload=extra_hw_frames=64");
             }
