diff -rupN sofia-sip-1.12.11.orig/libsofia-sip-ua/su/sofia-sip/su_os_nw.h sofia-sip-1.12.11/libsofia-sip-ua/su/sofia-sip/su_os_nw.h
--- sofia-sip-1.12.11.orig/libsofia-sip-ua/su/sofia-sip/su_os_nw.h	2021-07-02 13:29:58.390694642 +0200
+++ sofia-sip-1.12.11/libsofia-sip-ua/su/sofia-sip/su_os_nw.h	2021-07-02 13:31:06.167276581 +0200
@@ -58,7 +58,7 @@
 #endif
 
 #if SU_HAVE_POLL
-#include <sys/poll.h>
+#include <poll.h>
 #endif
 
 SOFIA_BEGIN_DECLS
diff -rupN sofia-sip-1.12.11.orig/libsofia-sip-ua/su/sofia-sip/su_wait.h sofia-sip-1.12.11/libsofia-sip-ua/su/sofia-sip/su_wait.h
--- sofia-sip-1.12.11.orig/libsofia-sip-ua/su/sofia-sip/su_wait.h	2021-07-02 13:29:58.390694642 +0200
+++ sofia-sip-1.12.11/libsofia-sip-ua/su/sofia-sip/su_wait.h	2021-07-02 13:31:22.943899439 +0200
@@ -47,7 +47,7 @@
 #endif
 
 #if SU_HAVE_POLL
-#include <sys/poll.h>
+#include <poll.h>
 #endif
 
 SOFIA_BEGIN_DECLS
